gradio/route_utils.py

```diff
@@ -583,3 +584,40 @@ def starts_with_protocol(string: str) -> bool:
     """
     pattern = r"^[a-zA-Z][a-zA-Z0-9+\-.]*://"
     return re.match(pattern, string) is not None
+
+
+class CustomCORSMiddleware(BaseHTTPMiddleware):
+    async def dispatch(self, request: fastapi.Request, call_next):
+        # Extract host and origin from the request headers
+        host_header: str = request.headers.get("host", "")
+        origin_header: str = (
+            request.headers.get("origin", "")
+            .replace("http://", "")
+            .replace("https://", "")
+        )
+
+        localhost_prefixes = ("localhost", "127.0.0.1", "0.0.0.0")
+
+        # Check if the host starts with localhost and origin is not localhost
+        if (
+            host_header.startswith(localhost_prefixes)
+            and origin_header
+            and not origin_header.startswith(localhost_prefixes)
+        ):
+            # Block the request by returning a 403 Forbidden response
+            return fastapi.Response(
+                content="Blocked by CORS policy.",
+                status_code=fastapi.status.HTTP_403_FORBIDDEN,
+            )
+
+        response = await call_next(request)
+        response.headers["Access-Control-Allow-Origin"] = request.headers.get(
+            "origin", "*"
+        )  # Allow the requesting origin
```

Comment:
So in case the request doesn't specify an origin, you reply with "all origins are fine"? That sounds suspect...

Reply:
Yes, we rely on the browser to pass in the origin header. If the request is coming from a non-browser setup, like cURL or the gradio python/js clients, this will be empty and we do want to allow that

Reply:
I think you shouldn't be responding with any CORS headers if there is no Origin header. Also, you may want to look at how the original CORS middleware handles e.g. preflight requests, and see if that's applicable here...

Reply:
Oh yeah I see what you mean I think you're right

Reply:
Actually thinking about this some more, I believe the original code is fine as it is. The origin header is set by the browser, so when it is not specified, the request must be coming from a command line environment, in which case, we are ok with serving all assets. Note that the original CORS policy (before this PR) was essentially this for all requests:

```py
            app.add_middleware(
                CORSMiddleware,
                allow_origins=["*"],
                allow_methods=["*"],
                allow_headers=["*"],
            )
```

Reply:
I've tightened this up -- now only the original origin, if present, is returned in the `access-control-allow-origin` header. If no origin is present, then no header is added.